### DISCUSSION TITLE
Is there any better documentation for how to do indentation queries?

### DISCUSSION THREAD

**Discussion by Hubro** [nvim-treesitter/nvim-treesitter#5259] [Q&A] [answered]
The only documentation on indent queries I can find is in CONTRIBUTING.md, which only contains:

```
@indent.begin       ; indent children when matching this node
@indent.end         ; marks the end of indented block
@indent.align       ; behaves like python aligned/hanging indent
@indent.dedent      ; dedent children when matching this node
@indent.branch      ; dedent itself when matching this node
@indent.ignore      ; do not indent in this node
@indent.auto        ; behaves like 'autoindent' buffer option
@indent.zero        ; sets this node at position 0 (no indent)
```

Is this it?

First of all I don't think the documentation here makes it clear enough how these captures work, they should all definitely have some usage examples that demonstrate their effect.

Second of all, this doesn't even document everything that can be done. Just by opening [`python/indents.scm`](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/python/indents.scm) I immediately see a bunch of undocumented things like:

- `(#set! indent.open_delimiter "[")`
- `(#set! indent.close_delimiter "]")`
- `(#set! indent.immediate 1)`
- `(#set! indent.avoid_last_matching_next 1)`
- `(#lua-match? @indent.branch "^else")`

And I wouldn't be surprised if there's a bunch more that can be done.

So I'm down to raw trial and error to try to figure out how each capture behaves and how they must be combined to make indentation work as expected.

As an added pain point, I don't have a good development cycle. I can use `tree-sitter query` to test my queries, but that doesn't do much good since I only have a vague idea what effect each capture has in the editor.

That means I basically have to edit the query on disk, quit my editor, reopen the editor on a relevant file and try some indent operations to see if my change worked. This is frustratingly slow. Is there no better way?

---

TL;DR

I'm finding writing indent queries a very painful experience so far, and I think (since this is a contribution-driven project) it would benefit massively from better documentation.

**Comment by clason** [comment, marked as answer]
Indentation is WIP, and we're still in the business of ironing things out, so the interface is not stable yet. 

Right now, the other indent queries are indeed the best guideline there is. You can also ask on Matrix.

>  (since this is a contribution-driven project) 

Yes, so contributions -- including to the documentation -- are welcome.

See also https://github.com/nvim-treesitter/nvim-treesitter/pull/4830

**Comment by michaelfortunato** [comment]
For what it's worth I am trying to add more Typst indent queries and feel the same way as OP.
